clients/src/main/java/org/apache/kafka/common/metrics/stats/SampledStat.java

```diff
@@ -106,44 +109,51 @@ public String toString() {
 
     public abstract double combine(List<Sample> samples, MetricConfig config, long now);
 
-    /* Timeout any windows that have expired in the absence of any events */
+    // purge any samples that lack observed events within the monitored window
     protected void purgeObsoleteSamples(MetricConfig config, long now) {
         long expireAge = config.samples() * config.timeWindowMs();
         for (Sample sample : samples) {
-            if (now - sample.lastWindowMs >= expireAge)
+            // samples overlapping the monitored window are kept,
+            // even if they started before it
+            if (now - sample.lastEventMs >= expireAge) {
                 sample.reset(now);
+            }
         }
     }
 
     protected static class Sample {
         public double initialValue;
         public long eventCount;
-        public long lastWindowMs;
+        public long startTimeMs;
+        public long lastEventMs;
         public double value;
 
         public Sample(double initialValue, long now) {
             this.initialValue = initialValue;
             this.eventCount = 0;
-            this.lastWindowMs = now;
+            this.startTimeMs = now;
+            this.lastEventMs = now;
             this.value = initialValue;
         }
 
         public void reset(long now) {
             this.eventCount = 0;
-            this.lastWindowMs = now;
+            this.startTimeMs = now;
+            this.lastEventMs = now;
             this.value = initialValue;
         }
 
         public boolean isComplete(long timeMs, MetricConfig config) {
-            return timeMs - lastWindowMs >= config.timeWindowMs() || eventCount >= config.eventWindow();
+            return timeMs - startTimeMs >= config.timeWindowMs() || eventCount >= config.eventWindow();
```

Comment:
> in this case, totalElapsedTimeMs will be equal to (config.samples() - 1) * timeWindowMs, and then we will get smaller measure due to the larger denominator

@chia7712: Could you explain your point a bit more? In the current design, we always ensure at least `(config.samples() - 1) * timeWindowMs` in `totalElapsedTimeMs`. This is to avoid the potential extreme large measured value when the accumulated sample window is small (e.g. when the metric is first recorded).

Reply:
Please ignore my comments. see following test
```java
    private double test(long eventWindow) {
        Rate rate = new Rate();
        Time timer = new MockTime();
        final MetricConfig config = new MetricConfig().samples(2).timeWindow(1, SECONDS).eventWindow(eventWindow);
        // generate 9 records
        IntStream.range(0, 9).forEach(__ -> {
            rate.record(config, 1, timer.milliseconds());
            timer.sleep(1);
        });
        // almost 1 second
        timer.sleep(990);
        return rate.measure(config, timer.milliseconds());
    }
```

**before**: 3 / second
**after**: 4.5 / second

I neglect that this approach has a extra sample, and so the measure value can be larger than before.

Reply:
Yes, when the oldest window overlaps with the monitored window, the old approach ignores it and the current approach keeps it. Either one could be considered correct.